### PR TITLE
8293121: (fs) Refactor UnixFileSystem copying into generic Unix, Linux, and BSD implementations

### PR DESCRIPTION
**Pull request by bplb**
Refactor `UnixFileSystem` and related classes to move implementations specific to Linux and BSD (macOS) into OS-specific classes and native code. Generic Unix implementations remain in existing `Unix*` files.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293121](https://bugs.openjdk.org/browse/JDK-8293121): (fs) Refactor UnixFileSystem copying into generic Unix, Linux, and BSD implementations

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10093/head:pull/10093` \
`$ git checkout pull/10093`

Update a local copy of the PR: \
`$ git checkout pull/10093` \
`$ git pull https://git.openjdk.org/jdk pull/10093/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10093`

View PR using the GUI difftool: \
`$ git pr show -t 10093`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10093.diff">https://git.openjdk.org/jdk/pull/10093.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
The refactoring means there is a 1-1 relationship between UnixFileSystem and a state-less UnixFileCopy.  If you just move the methods to UnixFileSystem then it will give you the framework to add Linux overrides to LinuxFileSystem and macOS/BSD overrides to  BsdFileSystem. That would avoid needing to create a parallel hierarchy of classes just to support copy/move.

**Comment by bplb** [comment]
> If you just move the methods to UnixFileSystem then it will give you the framework to add Linux overrides to LinuxFileSystem and macOS/BSD overrides to BsdFileSystem.

I was hesitant to add so much code to `UnixFileSystem` but it does seem logical.

**Comment by AlanBateman** [comment]
I skimmed through the changes in be36c5a6 and this is a good refactor. I don't have time to do a detailed walk through right now but I'll get to that soon.

File: src/java.base/linux/native/libnio/fs/LinuxNativeDispatcher.c [hunk 1]
```diff
@@ -139,6 +140,13 @@ JNIEXPORT void JNICALL
...
-    /* FIXME - man page doesn't explain how errors are returned */
+    // The endmntent() function always returns 1.
     endmntent(fp);
 }
+
+JNIEXPORT jint JNICALL
+Java_sun_nio_fs_LinuxNativeDispatcher_posix_1fadvise(JNIEnv* env, jclass this,
+    jint fd, jlong offset, jlong len, jint advice)
+{
+    return posix_fadvise((int)fd, (off_t)offset, (off_t)len, (int)advice);
```
Comment (AlanBateman) [inline src/java.base/linux/native/libnio/fs/LinuxNativeDispatcher.c:151]: I assume this should posix_fadvise64 as you'll need a off64_t on 32-bit systems.

File: test/jdk/java/nio/file/Files/CopyAndMove.java [hunk 2]
```diff
@@ -22,7 +22,7 @@
  */
 
 /* @test
- * @bug 4313887 6838333 6917021 7006126 6950237 8006645 8201407 8267820
+ * @bug 4313887 6838333 6917021 7006126 6950237 8006645 8201407 8293121 8267820
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/Files/CopyAndMove.java:25]: Did you mean to include this test update in this PR?

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java [hunk 3]
```diff
@@ -355,4 +380,703 @@ String normalizeNativePath(String path) {
...
+                int factor = (MIN_BUFFER_SIZE + bufferSize - 1)/bufferSize;
+                bufferSize *= factor;
+            }
+        } catch (UnixException ignored) {
+        }
+        return bufferSize;
+    }
+
+    // whether direct copying is supported on this platform
+    private static volatile boolean directCopyNotSupported;
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java:428]: Would it be possible to move MIN_BUFFER_SIZE and directCopyNotSupported to the top so they are declared before the instance fields?

**Review by AlanBateman — COMMENTED** [review]
Overall I think the structure looks quite good. A lot of code has moved so it's hard to see what has changed so I focused on the copyFile and supporting methods as that seems to be where the only changes are.

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java [hunk 4]
```diff
@@ -121,10 +127,56 @@ List<UnixMountEntry> getMountEntries() {
...
-
     @Override
     FileStore getFileStore(UnixMountEntry entry) throws IOException {
         return new LinuxFileStore(this, entry);
     }
+
+    // --- file copying ---
+
+    @Override
+    protected void bufferedCopy(int dst, int src, long address,
```
Comment (AlanBateman) [inline src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java:138]: Everything is in sun.nio.fs and the existing methods aren't protected. I don't mind if we have the protected methods but it's a bit inconsistent to have it on some but not all of the methods that are intended to be used by subclasses.

File: test/jdk/java/nio/file/Files/CopyAndMove.java (hunk 2 above)
Comment (bplb) [reply test/jdk/java/nio/file/Files/CopyAndMove.java:25]: I was thinking of the coverage but as this really adds a CoW file system if available then I think it better to leave it in the other PR about cloning on Linux.

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java (hunk 4 above)
Comment (bplb) [reply src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java:138]: I was thinking that `protected` would suggest that the method should likely be overridden.

File: src/java.base/linux/native/libnio/fs/LinuxNativeDispatcher.c (hunk 1 above)
Comment (bplb) [reply src/java.base/linux/native/libnio/fs/LinuxNativeDispatcher.c]: Fixed in b6cb94901d1bbaa34917a5aaf381ecf700282c54.

File: test/jdk/java/nio/file/Files/CopyAndMove.java (hunk 2 above)
Comment (bplb) [reply test/jdk/java/nio/file/Files/CopyAndMove.java]: Removed in b6cb94901d1bbaa34917a5aaf381ecf700282c54.

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java (hunk 4 above)
Comment (bplb) [reply src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java]: Removed `protected` in b6cb94901d1bbaa34917a5aaf381ecf700282c54.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java (hunk 3 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java]: Fixed in b6cb94901d1bbaa34917a5aaf381ecf700282c54.

**Review by AlanBateman — APPROVED** [review]
Thanks for the updates, I don't have any comments/issues.

**Comment by bplb** [comment]
> Thanks for the updates, I don't have any comments/issues.

Thanks for the review!

**Pull request closed without merging**